src/compiler/core.ts

```diff
@@ -2144,6 +2182,23 @@ export function memoizeOne<A extends string | number | boolean | undefined, T>(c
     };
 }
 
+/**
+ * A version of `memoize` that supports a single non-primitive argument, stored as keys of a WeakMap.
+ *
+ * @internal
+ */
+export function memoizeWeak<A extends object, T>(callback: (arg: A) => T): (arg: A) => T {
+    const map = new WeakMap<A, T>();

```

Comment:
Honestly, I'm curious if there's any reason to _not_ use a `WeakMap` in `memoizeOne` by default; are `WeakMaps` actually that much slower than regular ones? `memoizeOne` is only used in one place, but that place doesn't seem like it'd benefit from weakness anyway (I was hoping there would be some secret memory win).

Reply:
WeakMaps can’t take primitive keys. I tested performance on non-primitives in Chrome with the following code and the performance was nearly identical:

```js
// setup

function memoizeMap(callback) {
	const map = new Map();
    return (arg) => {
        let value = map.get(arg);
        if (value === undefined && !map.has(arg)) {
            value = callback(arg);
            map.set(arg, value);
        }
        return value;
    };
}

function memoizeWeakMap(callback) {
	const map = new WeakMap();
    return (arg) => {
        let value = map.get(arg);
        if (value === undefined && !map.has(arg)) {
            value = callback(arg);
            map.set(arg, value);
        }
        return value;
    };
}

const objects = [
  ...Object.values(window).filter(v => v && (typeof v === 'object' || typeof v === 'function')),
  ...document.querySelectorAll('*')
];

function test(memo) {
	const f = memo(arg => arg);
	for (const o of objects) {
		f(o);
		f(o);
	}
	for (const o of objects) {
		f(o);
	}
}

// case 1

test(memoizeMap)

// case 2

test(memoizeWeakMap)
```

Reply:
Shoot, yep, forgot about that.